src/jdk.internal.le/share/classes/jdk/internal/console/SimpleConsoleReader.java

```diff
@@ -0,0 +1,324 @@
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.console;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.io.Writer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.IntSupplier;
+import jdk.internal.org.jline.utils.WCWidth;
+
+public class SimpleConsoleReader {
+
+    //public, to simplify access from tests:
+    public static char[] doRead(Reader reader, Writer out, boolean password, int firstLineOffset, IntSupplier terminalWidthSupplier) throws IOException {
+        CleanableBuffer result = new CleanableBuffer();
+        try {
+            doReadImpl(reader, out, password, firstLineOffset, terminalWidthSupplier, result);
+            return result.getData();
+        } finally {
+            result.zeroOut();
+        }
+    }
+
+    private static void doReadImpl(Reader reader, Writer out, boolean password, int firstLineOffset, IntSupplier terminalWidthSupplier, CleanableBuffer result) throws IOException {
+        int caret = 0;
+        int r;
+        PaintState prevState = new PaintState();
+
+        READ: while (true) {
+            //paint:
+            if (firstLineOffset != (-1) && !password) {
+                prevState = repaint(out, firstLineOffset, terminalWidthSupplier, result.data, result.length, caret, prevState);
+            }
+
+            //read
+            r = reader.read();
+            switch (r) {
+                case -1: continue READ;
+                case '\n', '\r': break READ;
+                case 4: break READ; //EOF/Ctrl-D
+                case 127:
+                    //backspace:
+                    if (caret > 0) {
+                        result.delete(caret - 1, caret);
+                        caret--;
+                    }
+                    continue READ;
+                case '\033':
```

Comment:
If this is meant to be platform-agnostic, is it really safe to make these assumptions about the ability to produce or interpret escape codes and to make assumptions about their behavior on the user's terminal? I don't think it would even be safe to assume a single terminal type or interpretation on POSIX-type OSes; that's what things like `terminfo`/`termcap` are supposed to be for, right?

Reply:
In theory, yes, that's what `terminfo` is supposed to do. But `terminfo` has its own set of problems, and what we get from a terminal and what we send to the terminal is so simple and generally supported, that it should be possible to make it work on all realistic terminals(*). Or is there a particular terminal you have in mind?

(*) I admit I've forgot that `\E[7~`/`\E[8~` is Home/End as well. I'll fix that.

Reply:
I suppose I don't have a particular terminal in mind. But at the very least, maybe some documentation somewhere should reference e.g. [ECMA-48](https://ecma-international.org/publications-and-standards/standards/ecma-48/) as the basis for the behaviors of the console?